src/v/raft/replicate_entries_stm.cc

```diff
@@ -381,9 +381,48 @@ replicate_entries_stm::wait_for_majority_flush() {
     }
 }
 
+ss::future<result<replicate_result>>
+replicate_entries_stm::wait_for_majority_no_flush() {
+    if (!_append_result) {
+        co_return build_replicate_result();
+    }
+    auto appended_offset = _append_result->value().last_offset;
+    auto appended_term = _append_result->value().last_term;
+
+    auto stop_cond = [this, appended_offset, appended_term] {
+        const auto current_committed_offset = _ptr->committed_offset();
+        const auto current_majority_replicated
+          = _ptr->majority_replicated_index();
+        const auto replicated = current_majority_replicated >= appended_offset;
+        const auto truncated = _ptr->term() > appended_term
+                               && current_committed_offset
+                                    > _initial_committed_offset
+                               && _ptr->_log->get_term(appended_offset)
+                                    != appended_term;
+
+        return replicated || truncated;
+    };
+    try {
+        co_await _ptr->_majority_replicated_index_updated.wait(stop_cond);
+        co_return process_result(appended_offset, appended_term);
+    } catch (const ss::broken_condition_variable&) {
+        vlog(
+          _ctxlog.debug,
+          "Replication of entries with last offset: {}, flush: {} aborted - "
+          "shutting down",
+          _dirty_offset,
+          _is_flush_required);
+        co_return result<replicate_result>(
+          make_error_code(errc::shutting_down));
+    }
+}
+
```

Comment:
nit: this differs from `wait_for_majority_flush` only in the `stop_cond`. Maybe we can have a generic method that would accept the predicate as an argument ?